File: bindings/go/bridge.go
package innovate

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
)

const kernelOperationDiscoverModels = "discover_models"
const kernelOperationFitModel = "fit_model"
const kernelOperationPredictModel = "predict_model"
const kernelOperationSimulateModel = "simulate_model"
const kernelOperationSummarizeModel = "summarize_model"
const kernelOperationDiagnoseModel = "diagnose_model"

// KernelJSONValue represents a JSON-compatible kernel payload value.
type KernelJSONValue = any

// KernelOperation names a stable kernel operation.
type KernelOperation string

// KernelBridgeError mirrors the stable error envelope returned by the kernel bridge.
type KernelBridgeError struct {
	Code      string
	Message   string
	Operation string
	Details   map[string]KernelJSONValue
	Retryable bool
	Response  map[string]KernelJSONValue
}

func (e *KernelBridgeError) Error() string {
	return fmt.Sprintf("kernel bridge error (%s) for operation %q: %s", e.Code, e.Operation, e.Message)
}

// KernelDiscoveryRecord describes a discoverable model family.
type KernelDiscoveryRecord struct {
	Key                        string   `json:"key"`
	Family                     string   `json:"family"`
	ImportPath                 string   `json:"import_path"`
	Stability                  string   `json:"stability"`
	SupportsCovariates         bool     `json:"supports_covariates"`
	SupportsMultivariateOutput bool     `json:"supports_multivariate_output"`
	SupportedBackends          []string `json:"supported_backends"`
	OptionalDependencies       []string `json:"optional_dependencies,omitempty"`
	SupportsSimulation         bool     `json:"supports_simulation"`
	SupportsSummarize          bool     `json:"supports_summarize"`
}

// KernelDiscoveryResponse contains the discoverable model registry and bridge metadata.
type KernelDiscoveryResponse struct {
	SchemaVersion string                     `json:"schema_version"`
	Models        []KernelDiscoveryRecord    `json:"models"`
	Metadata      map[string]KernelJSONValue `json:"metadata,omitempty"`
}

// KernelPredictionResponse contains a normalized prediction or simulation result.
type KernelPredictionResponse struct {
	SchemaVersion string                     `json:"schema_version"`
	Operation     string                     `json:"operation"`
	ModelKey      string                     `json:"model_key"`
	ModelName     string                     `json:"model_name"`
	Family        string                     `json:"family"`
	Value         KernelJSONValue            `json:"value"`
	Metadata      map[string]KernelJSONValue `json:"metadata,omitempty"`
}

// KernelFitResponse contains the fitted model payload and diagnostics.
type KernelFitResponse struct {
	SchemaVersion string                     `json:"schema_version"`
	Operation     string                     `json:"operation"`
	ModelKey      string                     `json:"model_key"`
	ModelName     string                     `json:"model_name"`
	Family        string                     `json:"family"`
	Parameters    map[string]KernelJSONValue `json:"parameters"`
	Predictions   KernelJSONValue            `json:"predictions"`
	Diagnostics   map[string]KernelJSONValue `json:"diagnostics,omitempty"`
	State         map[string]KernelJSONValue `json:"state"`
	Metadata      map[string]KernelJSONValue `json:"metadata,omitempty"`
}

// KernelSummaryResponse contains a summarized fitted model payload.
type KernelSummaryResponse struct {
	SchemaVersion     string                     `json:"schema_version"`
	Operation         string                     `json:"operation"`
	ModelKey          string                     `json:"model_key"`
	ModelName         string                     `json:"model_name"`
	Family            string                     `json:"family"`
	ParameterNames    []string                   `json:"parameter_names"`
	Parameters        map[string]KernelJSONValue `json:"parameters"`
	ConstructorKwargs map[string]KernelJSONValue `json:"constructor_kwargs"`
	State             map[string]KernelJSONValue `json:"state"`
	Diagnostics       map[string]KernelJSONValue `json:"diagnostics,omitempty"`
	Metadata          map[string]KernelJSONValue `json:"metadata,omitempty"`
}

// KernelDiagnoseResponse contains a structured diagnostics contract for a fitted model.
type KernelDiagnoseResponse struct {
	SchemaVersion string                     `json:"schema_version"`
	Operation     string                     `json:"operation"`
	Diagnostics   map[string]KernelJSONValue `json:"diagnostics"`
	State         map[string]KernelJSONValue `json:"state"`
	Metadata      map[string]KernelJSONValue `json:"metadata,omitempty"`
}

func repoRoot() string {
	return filepath.Clean(filepath.Join(ModuleRoot(), "..", ".."))
}

func kernelCommandParts() ([]string, error) {
	command := strings.TrimSpace(os.Getenv("INNOVATE_PYTHON_COMMAND"))
	if command != "" {
		parts := strings.Fields(command)
		if len(parts) == 0 {
			return nil, errors.New("INNOVATE_PYTHON_COMMAND must not be empty")
		}
		return parts, nil
	}

	if _, err := exec.LookPath("uv"); err == nil {
		return []string{"uv", "run", "python"}, nil
	}
	if _, err := exec.LookPath("python3"); err == nil {
		return []string{"python3"}, nil
	}
	return nil, errors.New("unable to locate a Python launcher for the kernel bridge")
}

func kernelInvokeBridge(request KernelRequest) (map[string]KernelJSONValue, error) {
	tempDir, err := os.MkdirTemp("", "innovate-go-kernel-")
	if err != nil {
		return nil, err
	}
	defer os.RemoveAll(tempDir)

	requestPath := filepath.Join(tempDir, "request.json")
	responsePath := filepath.Join(tempDir, "response.json")

	requestBytes, err := json.MarshalIndent(request, "", "  ")
	if err != nil {
		return nil, err
	}
	if err := os.WriteFile(requestPath, append(requestBytes, '\n'), 0o600); err != nil {
		return nil, err
	}

	commandParts, err := kernelCommandParts()
	if err != nil {
		return nil, err
	}

	args := append([]string{}, commandParts[1:]...)
	args = append(args, BridgeScriptPath(), requestPath, responsePath)
	cmd := exec.Command(commandParts[0], args...)
	cmd.Dir = repoRoot()
	cmd.Env = append(os.Environ(), "PYTHONPATH="+strings.Join([]string{filepath.Join(repoRoot(), "src"), os.Getenv("PYTHONPATH")}, string(os.PathListSeparator)))

	output, runErr := cmd.CombinedOutput()
	if runErr != nil {
		return nil, &KernelBridgeError{
			Code:      "internal_error",
			Message:   "Kernel bridge process failed",
			Operation: request.Operation,
			Details: map[string]KernelJSONValue{
				"stdout": strings.TrimSpace(string(output)),
			},
			Retryable: false,
			Response:  map[string]KernelJSONValue{},
		}
	}

	responseBytes, err := os.ReadFile(responsePath)
	if err != nil {
		return nil, &KernelBridgeError{
			Code:      "internal_error",
			Message:   "Kernel bridge did not produce a response",
			Operation: request.Operation,
			Details: map[string]KernelJSONValue{
				"stdout": strings.TrimSpace(string(output)),
			},
			Retryable: false,
			Response:  map[string]KernelJSONValue{},
		}
	}

	var response map[string]KernelJSONValue
	if err := json.Unmarshal(responseBytes, &response); err != nil {
		return nil, err
	}

	if bridgeErr := kernelBridgeErrorFromResponse(response, request.Operation); bridgeErr != nil {
		return nil, bridgeErr
	}

	return response, nil
}

func kernelBridgeErrorFromResponse(response map[string]KernelJSONValue, fallbackOperation string) *KernelBridgeError {
	rawError, ok := response["error"].(map[string]any)
	if !ok || rawError == nil {
		return nil
	}

	operation := fallbackOperation
	if value, ok := rawError["operation"].(string); ok && value != "" {
		operation = value
	}

	details := map[string]KernelJSONValue{}
	if rawDetails, ok := rawError["details"].(map[string]any); ok && rawDetails != nil {
		details = normalizeMap(rawDetails)
	}

	retryable := false
	if value, ok := rawError["retryable"].(bool); ok {
		retryable = value
	}

	code, _ := rawError["code"].(string)
	message, _ := rawError["message"].(string)

	return &KernelBridgeError{
		Code:      code,
		Message:   message,
		Operation: operation,
		Details:   details,
		Retryable: retryable,
		Response:  normalizeMap(response),
	}
}

func normalizeKernelValue(value any) KernelJSONValue {
	switch typed := value.(type) {
	case nil, string, bool:
		return typed
	case float32:
		return float64(typed)
	case float64, int, int32, int64, uint32, uint64:
		return typed
	case []any:
		values := make([]KernelJSONValue, len(typed))
		for index, item := range typed {
			values[index] = normalizeKernelValue(item)
		}
		return values
	case map[string]any:
		if shape, ok := kernelShapeFromValue(typed["shape"]); ok {
			if values, ok := typed["values"].([]any); ok {
				normalized := make([]KernelJSONValue, len(values))
				for index, item := range values {
					normalized[index] = normalizeKernelValue(item)
				}
				return reshapeKernelArray(normalized, shape)
			}
		}

		if columns, ok := typed["columns"].([]any); ok {
			if rows, ok := typed["rows"].([]any); ok {
				columnNames := make([]string, 0, len(columns))
				for _, column := range columns {
					columnNames = append(columnNames, fmt.Sprint(column))
				}
				table := make([]map[string]KernelJSONValue, 0, len(rows))
				for _, rowValue := range rows {
					row, _ := rowValue.([]any)
					entry := make(map[string]KernelJSONValue, len(columnNames))
					for index, column := range columnNames {
						if index < len(row) {
							entry[column] = normalizeKernelValue(row[index])
						} else {
							entry[column] = nil
						}
					}
					table = append(table, entry)
				}
				return table
			}
		}

		return normalizeMap(typed)
	default:
		return typed
	}
}

func normalizeMap(values map[string]any) map[string]KernelJSONValue {
	normalized := make(map[string]KernelJSONValue, len(values))
	for key, value := range values {
		normalized[key] = normalizeKernelValue(value)
	}
	return normalized
}

func kernelShapeFromValue(value any) ([]int, bool) {
	rawShape, ok := value.([]any)
	if !ok {
		return nil, false
	}

	shape := make([]int, 0, len(rawShape))
	for _, item := range rawShape {
		switch dimension := item.(type) {
		case float64:
			shape = append(shape, int(dimension))
		case float32:
			shape = append(shape, int(dimension))
		case int:
			shape = append(shape, dimension)
		case int64:
			shape = append(shape, int(dimension))
		case uint64:
			shape = append(shape, int(dimension))
		case string:
			value, err := strconv.Atoi(dimension)
			if err != nil {
				return nil, false
			}
			shape = append(shape, value)
		default:
			return nil, false
		}
	}
	return shape, true
}

func reshapeKernelArray(values []KernelJSONValue, shape []int) KernelJSONValue {
	if len(shape) == 0 {
		if len(values) == 0 {
			return nil
		}
		return values[0]
	}

	dimension := shape[0]
	if dimension == 0 {
		return []KernelJSONValue{}
	}

	if len(shape) == 1 {
		end := dimension
		if end > len(values) {
			end = len(values)
		}
		return values[:end]
	}

	chunkSize := 1
	for _, item := range shape[1:] {
		chunkSize *= item
	}

	chunks := make([]KernelJSONValue, 0, dimension)
	for index := 0; index < dimension; index++ {
		start := index * chunkSize
		end := start + chunkSize
		if start >= len(values) {
			chunks = append(chunks, reshapeKernelArray(nil, shape[1:]))
			continue
		}
		if end > len(values) {
			end = len(values)
		}
		chunks = append(chunks, reshapeKernelArray(values[start:end], shape[1:]))
	}

	return chunks
}

func discoveryResponseFromMap(response map[string]KernelJSONValue) (KernelDiscoveryResponse, error) {
	rawModels, ok := response["models"].([]any)
	if !ok {
		return KernelDiscoveryResponse{}, errors.New("kernel discovery response must include models")
	}

	models := make([]KernelDiscoveryRecord, 0, len(rawModels))
	for _, rawModel := range rawModels {
		record, ok := rawModel.(map[string]any)
		if !ok {
			return KernelDiscoveryResponse{}, errors.New("kernel discovery records must be objects")
		}

		model := KernelDiscoveryRecord{
			Key:                        fmt.Sprint(record["key"]),
			Family:                     fmt.Sprint(record["family"]),
			ImportPath:                 fmt.Sprint(record["import_path"]),
			Stability:                  fmt.Sprint(record["stability"]),
			SupportsCovariates:         asBool(record["supports_covariates"]),
			SupportsMultivariateOutput: asBool(record["supports_multivariate_output"]),
			SupportedBackends:          stringSlice(record["supported_backends"]),
			OptionalDependencies:       stringSlice(record["optional_dependencies"]),
			SupportsSimulation:         asBool(record["supports_simulation"]),
			SupportsSummarize:          asBool(record["supports_summarize"]),
		}
		models = append(models, model)
	}

	return KernelDiscoveryResponse{
		SchemaVersion: fmt.Sprint(response["schema_version"]),
		Models:        models,
		Metadata:      mapFromKernelValue(response["metadata"]),
	}, nil
}

func mapFromKernelValue(value KernelJSONValue) map[string]KernelJSONValue {
	if value == nil {
		return map[string]KernelJSONValue{}
	}
	typed, ok := value.(map[string]KernelJSONValue)
	if ok {
		return typed
	}
	if raw, ok := value.(map[string]any); ok {
		return normalizeMap(raw)
	}
	return map[string]KernelJSONValue{}
}

func stringSlice(value KernelJSONValue) []string {
	raw, ok := value.([]any)
	if !ok {
		return nil
	}
	values := make([]string, 0, len(raw))
	for _, item := range raw {
		values = append(values, fmt.Sprint(item))
	}
	return values
}

func asBool(value any) bool {
	typed, ok := value.(bool)
	return ok && typed
}

// ExtractDiagnostics returns the diagnostics payload from a fitted, summarized, or diagnosed result.
func ExtractDiagnostics(result any) (map[string]KernelJSONValue, bool) {
	switch typed := result.(type) {
	case KernelFitResponse:
		return typed.Diagnostics, typed.Diagnostics != nil
	case *KernelFitResponse:
		if typed == nil {
			return nil, false
		}
		return typed.Diagnostics, typed.Diagnostics != nil
	case KernelSummaryResponse:
		return typed.Diagnostics, typed.Diagnostics != nil
	case *KernelSummaryResponse:
		if typed == nil {
			return nil, false
		}
		return typed.Diagnostics, typed.Diagnostics != nil
	case KernelDiagnoseResponse:
		return typed.Diagnostics, typed.Diagnostics != nil
	case *KernelDiagnoseResponse:
		if typed == nil {
			return nil, false
		}
		return typed.Diagnostics, typed.Diagnostics != nil
	case map[string]KernelJSONValue:
		if diagnostics, ok := typed["diagnostics"]; ok {
			return mapFromKernelValue(diagnostics), true
		}
	}

	return nil, false
}

func kernelResultEnvelope(response map[string]KernelJSONValue) (map[string]KernelJSONValue, error) {
	rawResult, ok := response["result"]
	if !ok || rawResult == nil {
		return map[string]KernelJSONValue{}, errors.New("kernel response did not include a result")
	}

	result, ok := rawResult.(map[string]any)
	if !ok {
		return map[string]KernelJSONValue{"value": normalizeKernelValue(rawResult)}, nil
	}

	return normalizeMap(result), nil
}

// DiscoverModels returns the registry of discoverable kernel models.
func DiscoverModels() (KernelDiscoveryResponse, error) {
	response, err := kernelInvokeBridge(KernelRequest{
		SchemaVersion: KernelSchemaVersion(),
		Operation:     kernelOperationDiscoverModels,
		ModelKey:      nil,
		Payload:       map[string]any{},
		Metadata:      map[string]any{},
	})
	if err != nil {
		return KernelDiscoveryResponse{}, err
	}

	return discoveryResponseFromMap(response)
}

// FitModel fits a stable model using the shared Python kernel.
func FitModel(request KernelRequest) (KernelFitResponse, error) {
	response, err := kernelInvokeBridge(request)
	if err != nil {
		return KernelFitResponse{}, err
	}

	result, err := kernelResultEnvelope(response)
	if err != nil {
		return KernelFitResponse{}, err
	}

	return KernelFitResponse{
		SchemaVersion: fmt.Sprint(response["schema_version"]),
		Operation:     fmt.Sprint(response["operation"]),
		ModelKey:      fmt.Sprint(result["model_key"]),
		ModelName:     fmt.Sprint(result["model_name"]),
		Family:        fmt.Sprint(result["family"]),
		Parameters:    mapFromKernelValue(result["parameters"]),
		Predictions:   normalizeKernelValue(result["predictions"]),
		Diagnostics:   mapFromKernelValue(result["diagnostics"]),
		State:         mapFromKernelValue(result["state"]),
		Metadata:      mapFromKernelValue(response["metadata"]),
	}, nil
}

// PredictModel predicts from a fitted stable model.
func PredictModel(request KernelRequest) (KernelPredictionResponse, error) {
	response, err := kernelInvokeBridge(request)
	if err != nil {
		return KernelPredictionResponse{}, err
	}

	metadata := mapFromKernelValue(response["metadata"])
	return KernelPredictionResponse{
		SchemaVersion: fmt.Sprint(response["schema_version"]),
		Operation:     fmt.Sprint(response["operation"]),
		ModelKey:      fmt.Sprint(metadata["model_key"]),
		ModelName:     fmt.Sprint(metadata["model_name"]),
		Family:        fmt.Sprint(metadata["family"]),
		Value:         normalizeKernelValue(response["result"]),
		Metadata:      metadata,
	}, nil
}

// SimulateModel simulates a fitted stable model.
func SimulateModel(request KernelRequest) (KernelPredictionResponse, error) {
	response, err := kernelInvokeBridge(request)
	if err != nil {
		return KernelPredictionResponse{}, err
	}

	metadata := mapFromKernelValue(response["metadata"])
	return KernelPredictionResponse{
		SchemaVersion: fmt.Sprint(response["schema_version"]),
		Operation:     fmt.Sprint(response["operation"]),
		ModelKey:      fmt.Sprint(metadata["model_key"]),
		ModelName:     fmt.Sprint(metadata["model_name"]),
		Family:        fmt.Sprint(metadata["family"]),
		Value:         normalizeKernelValue(response["result"]),
		Metadata:      metadata,
	}, nil
}

// SummarizeModel summarizes a fitted stable model.
func SummarizeModel(request KernelRequest) (KernelSummaryResponse, error) {
	response, err := kernelInvokeBridge(request)
	if err != nil {
		return KernelSummaryResponse{}, err
	}

	result, err := kernelResultEnvelope(response)
	if err != nil {
		return KernelSummaryResponse{}, err
	}

	return KernelSummaryResponse{
		SchemaVersion:     fmt.Sprint(response["schema_version"]),
		Operation:         fmt.Sprint(response["operation"]),
		ModelKey:          fmt.Sprint(result["model_key"]),
		ModelName:         fmt.Sprint(result["model_name"]),
		Family:            fmt.Sprint(result["family"]),
		ParameterNames:    stringSlice(result["parameter_names"]),
		Parameters:        mapFromKernelValue(result["parameters"]),
		ConstructorKwargs: mapFromKernelValue(result["constructor_kwargs"]),
		State:             mapFromKernelValue(result["state"]),
		Diagnostics:       mapFromKernelValue(result["diagnostics"]),
		Metadata:          mapFromKernelValue(response["metadata"]),
	}, nil
}

// DiagnoseModel returns a structured diagnostics contract for a fitted stable model.
func DiagnoseModel(request KernelRequest) (KernelDiagnoseResponse, error) {
	response, err := kernelInvokeBridge(request)
	if err != nil {
		return KernelDiagnoseResponse{}, err
	}

	result, err := kernelResultEnvelope(response)
	if err != nil {
		return KernelDiagnoseResponse{}, err
	}

	return KernelDiagnoseResponse{
		SchemaVersion: fmt.Sprint(response["schema_version"]),
		Operation:     fmt.Sprint(response["operation"]),
		Diagnostics:   mapFromKernelValue(result["diagnostics"]),
		State:         mapFromKernelValue(result["state"]),
		Metadata:      mapFromKernelValue(response["metadata"]),
	}, nil
}

File: bindings/typescript/src/bridge.ts
import { spawnSync } from "node:child_process";
import { existsSync, mkdtempSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { delimiter, join } from "node:path";
import { tmpdir } from "node:os";

import {
  kernelBridgeScript,
  kernelPackageRoot,
  kernelPythonCommand,
  kernelRepoRootOrNull,
  kernelRequest,
  type KernelRequest,
} from "./kernel.js";

export type KernelJSONValue =
  | string
  | number
  | boolean
  | null
  | KernelJSONValue[]
  | { [key: string]: KernelJSONValue };

export interface KernelDiscoveryRecord {
  key: string;
  family: string;
  import_path: string;
  stability: string;
  supports_covariates: boolean;
  supports_multivariate_output: boolean;
  supported_backends: string[];
  optional_dependencies: string[];
  supports_simulation: boolean;
  supports_summarize: boolean;
}

export interface KernelBridgeErrorDetails {
  [key: string]: KernelJSONValue;
}

export interface KernelBridgeErrorResponse {
  code: string;
  message: string;
  operation?: string | null;
  details?: KernelBridgeErrorDetails;
  retryable?: boolean;
}

export interface KernelFitResponse {
  model_key: string;
  model_name: string;
  family: string;
  parameters: Record<string, KernelJSONValue>;
  predictions: KernelJSONValue;
  diagnostics?: Record<string, KernelJSONValue>;
  state: Record<string, KernelJSONValue>;
}

export interface KernelSummaryResponse {
  model_key: string;
  model_name: string;
  family: string;
  parameter_names: string[];
  parameters: Record<string, KernelJSONValue>;
  constructor_kwargs: Record<string, KernelJSONValue>;
  state: Record<string, KernelJSONValue>;
  diagnostics?: Record<string, KernelJSONValue>;
}

export interface KernelDiagnoseResponse {
  diagnostics: Record<string, KernelJSONValue>;
  state: Record<string, KernelJSONValue>;
}

function splitCommand(command: string): string[] {
  const parts = command.trim().split(/\s+/).filter(Boolean);
  if (parts.length === 0) {
    throw new Error("INNOVATE_PYTHON_COMMAND must not be empty");
  }
  return parts;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function normalizeKernelValue(value: unknown): KernelJSONValue {
  if (
    value === null ||
    typeof value === "string" ||
    typeof value === "number" ||
    typeof value === "boolean"
  ) {
    return value;
  }

  if (Array.isArray(value)) {
    return value.map((item) => normalizeKernelValue(item));
  }

  if (isRecord(value)) {
    if (
      Array.isArray(value.shape) &&
      value.shape.every((dimension) => typeof dimension === "number") &&
      Array.isArray(value.values)
    ) {
      const values = value.values.map((item) => normalizeKernelValue(item));
      return reshapeKernelArray(values, value.shape as number[]);
    }

    if (
      Array.isArray(value.columns) &&
      Array.isArray(value.rows) &&
      value.rows.every((row) => Array.isArray(row))
    ) {
      const columns = value.columns.map((column) => String(column));
      return value.rows.map((row) => {
        const entries = row as unknown[];
        return Object.fromEntries(
          columns.map((column, index) => [column, normalizeKernelValue(entries[index])]),
        );
      });
    }

    return Object.fromEntries(
      Object.entries(value).map(([key, item]) => [key, normalizeKernelValue(item)]),
    );
  }

  return value as never;
}

function reshapeKernelArray(values: KernelJSONValue[], shape: number[]): KernelJSONValue {
  if (shape.length === 0) {
    return values[0] ?? null;
  }

  const [dimension, ...rest] = shape;
  if (dimension === 0) {
    return [];
  }

  if (rest.length === 0) {
    return values.slice(0, dimension);
  }

  const chunkSize = rest.reduce((product, item) => product * item, 1);
  const chunks: KernelJSONValue[] = [];
  for (let index = 0; index < dimension; index += 1) {
    const start = index * chunkSize;
    const end = start + chunkSize;
    chunks.push(reshapeKernelArray(values.slice(start, end), rest));
  }
  return chunks;
}

function responseError(response: Record<string, unknown>): KernelBridgeError | null {
  if (!isRecord(response.error) || response.error == null) {
    return null;
  }

  return KernelBridgeError.fromResponse(response, response.operation);
}

function decodeBridgeResponse(response: Record<string, unknown>): KernelJSONValue {
  const error = responseError(response);
  if (error) {
    throw error;
  }

  if (Array.isArray(response.models) && !("operation" in response)) {
    return response.models.map((model) => normalizeKernelValue(model));
  }

  if ("result" in response) {
    return normalizeKernelValue(response.result);
  }

  return normalizeKernelValue(response);
}

function invokeBridge(request: KernelRequest): KernelJSONValue {
  const tempDir = mkdtempSync(join(tmpdir(), "innovate-typescript-kernel-"));
  const requestPath = join(tempDir, "request.json");
  const responsePath = join(tempDir, "response.json");
  const command = splitCommand(kernelPythonCommand());
  const repoRoot = kernelRepoRootOrNull();
  const env = {
    ...process.env,
    PYTHONPATH: [repoRoot ? join(repoRoot, "src") : null, process.env.PYTHONPATH]
      .filter((part) => Boolean(part))
      .join(delimiter),
  };

  try {
    writeFileSync(requestPath, `${JSON.stringify(request, null, 2)}\n`, "utf8");
    const result = spawnSync(command[0], [...command.slice(1), kernelBridgeScript(), requestPath, responsePath], {
      cwd: repoRoot ?? kernelPackageRoot(),
      env,
      encoding: "utf8",
    });

    if (result.error) {
      throw new KernelBridgeError("Kernel bridge process failed", {
        code: "internal_error",
        operation: request.operation,
        details: { stderr: result.stderr?.toString() ?? "", stdout: result.stdout?.toString() ?? "" },
        retryable: false,
      });
    }

    if (!existsSync(responsePath)) {
      throw new KernelBridgeError("Kernel bridge did not produce a response", {
        code: "internal_error",
        operation: request.operation,
        details: { stderr: result.stderr?.toString() ?? "", stdout: result.stdout?.toString() ?? "" },
        retryable: false,
      });
    }

    const response = JSON.parse(readFileSync(responsePath, "utf8")) as Record<string, unknown>;
    return decodeBridgeResponse(response);
  } finally {
    rmSync(tempDir, { recursive: true, force: true });
  }
}

export class KernelBridgeError extends Error {
  readonly code: string;
  readonly operation: string;
  readonly details: KernelBridgeErrorDetails;
  readonly retryable: boolean;
  readonly response: Record<string, KernelJSONValue>;

  constructor(
    message: string,
    options: {
      code: string;
      operation: string;
      details?: KernelBridgeErrorDetails;
      retryable?: boolean;
      response?: Record<string, KernelJSONValue>;
    },
  ) {
    super(message);
    this.name = "KernelBridgeError";
    this.code = options.code;
    this.operation = options.operation;
    this.details = options.details ?? {};
    this.retryable = options.retryable ?? false;
    this.response = options.response ?? {};
    Object.setPrototypeOf(this, new.target.prototype);
  }

  static fromResponse(
    response: Record<string, unknown>,
    fallbackOperation?: unknown,
  ): KernelBridgeError {
    const error = isRecord(response.error) ? response.error : {};
    const operation =
      typeof error.operation === "string" && error.operation.length > 0
        ? error.operation
        : typeof fallbackOperation === "string" && fallbackOperation.length > 0
          ? fallbackOperation
          : "discover_models";
    const details = isRecord(error.details)
      ? Object.fromEntries(
          Object.entries(error.details).map(([key, value]) => [key, normalizeKernelValue(value)]),
        )
      : {};

    return new KernelBridgeError(String(error.message ?? "Kernel bridge error"), {
      code: String(error.code ?? "internal_error"),
      operation,
      details,
      retryable: Boolean(error.retryable ?? false),
      response: Object.fromEntries(
        Object.entries(response).map(([key, value]) => [key, normalizeKernelValue(value)]),
      ),
    });
  }
}

/** Discover the stable model registry through the kernel bridge. */
export function kernelDiscoverModels(): KernelDiscoveryRecord[] {
  return invokeBridge(kernelRequest("discover_models")) as unknown as KernelDiscoveryRecord[];
}

/** Fit a stable model through the kernel bridge. */
export function kernelFitModel(request: KernelRequest): KernelFitResponse {
  return invokeBridge(request) as unknown as KernelFitResponse;
}

/** Predict from a fitted stable model through the kernel bridge. */
export function kernelPredictModel(request: KernelRequest): KernelJSONValue {
  return invokeBridge(request);
}

/** Simulate a stable model through the kernel bridge. */
export function kernelSimulateModel(request: KernelRequest): KernelJSONValue {
  return invokeBridge(request);
}

/** Summarize a fitted stable model through the kernel bridge. */
export function kernelSummarizeModel(request: KernelRequest): KernelSummaryResponse {
  return invokeBridge(request) as unknown as KernelSummaryResponse;
}

/** Diagnose a fitted stable model through the kernel bridge. */
export function kernelDiagnoseModel(request: KernelRequest): KernelDiagnoseResponse {
  return invokeBridge(request) as unknown as KernelDiagnoseResponse;
}

/** Extract diagnostics from a kernel response payload when present. */
export function kernelExtractDiagnostics(
  result: unknown,
): Record<string, KernelJSONValue> | null {
  if (isRecord(result) && isRecord(result.diagnostics)) {
    return Object.fromEntries(
      Object.entries(result.diagnostics).map(([key, value]) => [key, normalizeKernelValue(value)]),
    );
  }

  return null;
}
